### DISCUSSION TITLE
Access data (user, accessToken, router, collections) in custom module

### DISCUSSION THREAD

**Discussion by chomamateusz** [directus/directus#4659] [Archive]
In custom module I should be on "blank canvas for creating anything you need" (at least that says documentation ;)) BUT how I can access the data? Is there a way to access:
1. user data and accessToken to act as the user
2. collections data
3. router to be able to change route without reloading page

I've tried in a lot of ways, and I have found:
1. We have user data in the argument of `preRegisterCheck` function that we can create in config, but no `accesToken`, and it is hacking this out. We can also do `/auth/refresh` with `"credentials": "include"` to use `refreshToken` from cookie and get new `accessToken` but it seems ultra hacky too :/ Why we can't just pass Directus SDK instance to the module config like `i18n` (https://github.com/directus/directus/blob/main/app/src/modules/define.ts#L8)?
2. With this access token I can set instance of Directus SDK or use REST API. If the Directus SDK instance would be available in the config this problem will disappear.
3. Can't figure out nothing except overwrite `window.location` and reload :/

**Comment by rijkvanzanten** [comment]
Heya!

The app `provide`-s a couple useful things, including a pre-authenticated `axios` instance you can use for all authenticated requests, and access to the internal stores, that contain all sorts of useful info on the schema/user:

https://github.com/directus/directus/blob/549fbd8b93e21649266758c9742be93a583b9121/app/src/app.vue#L111-L116

You can use this through Vue's provide/inject system as follows:

```js
export default {
  inject: ['system'],
  created() {
    // Read from the system stores
    const userStore = this.system.useUserStore();
    console.log(userStore.state);

    // Use the API
    this.system.api.get('/items/movies').then(res => console.log(res));
  }
}
```

**Reply by chomamateusz** [reply]
Wow! That is exactly that I need! THANKS!

1. It should work also in displays and interfaces the same way?
2. There are more things to be injected that way?

I'm not so good at Vue, so I tried to achieve it in pure JS ways as you can see. Even I did a template with React in Vue configured to be able to bring components that I already have (you probably already seen that https://github.com/amazingdesign/directus-interface--base) :)

**Reply by chomamateusz** [reply]
Ok I see the comment in the code 'This allows custom extensions to use the apps internals' so I've answered myself on 1. question and probably 2. too :)

**Reply by rijkvanzanten** [reply]
> It should work also in displays and interfaces the same way?

Yes! Same thing 🙂 

> There are more things to be injected that way?

As of right now, the (limited) list in that code snippet is what you get available inside of app extensions 👍🏻  Lemme know if you have any specifics you'd like to see added to that provider though!

**Reply by chomamateusz** [reply]
@rijkvanzanten do we have a TS typing for this `system` object that I could import in custom module?

**Reply by chomamateusz** [reply]
Another one, is there a way to use router?

**Reply by rijkvanzanten** [reply]
> @rijkvanzanten do we have a TS typing for this system object that I could import in custom module?

Hmm no don't think so. The snippet I sent above is all it does

> Another one, is there a way to use router?

Vue Router is globally available in `this.$router` (that's a Vue specific thing, not a Directus thing)

**Reply by dallashuggins** [reply]
Is this available for interfaces as well as modules? 

It allows me to import `stores`, `api`, `addTokenToURL`, and/or `axios` from Directus, but when I build the app I get many warnings, but it does finally build, but then when you start the app there's many "could not resolve" warnings related to importing these files. When opening the collection using the interface, the dev console specifies it cannot load custom interfaces and `TypeError: Error resolving module specifier “crypto”. Relative module specifiers must start with “./”, “../” or “/”.`. Building and starting works without importing these into the interface. I also tried using `system: true` when defining the interface with the same result (seen in the interface Typescript definition, not sure if what it's used for). 

Start of `defineComponent`:
```
export default defineComponent({
	props: {
		value: {
			type: String,
      default: null,
		},
	},
	emits: ['input'],
	inject: ['system'],
```

In setup:
```
provide('system', {
   ...stores, api, axios, addTokenToURL,
})
```

If this is available for interfaces, any ideas what I may be doing wrong? 

Thanks!

**Reply by rijkvanzanten** [reply]
@dallashuggins I believe this was since updated to rely on new composables to inject these shared properties. @nickrum what's the latest example for this?

**Reply by nickrum** [reply]
I don't think there is an example yet that is using the new composables. The only one we have right now is [this](https://docs.directus.io/guides/modules/#accessing-the-api-from-within-your-extension) one.

I added a more thorough answer to https://github.com/directus/directus/discussions/8534.

**Reply by dallashuggins** [reply]
Thank you for the quick response! Got the API working now. 🥳

**Reply by drennvinn** [reply]
@rijkvanzanten  Hi! The app/src/app.vue has changed since the last post so is it possible to get an update of how to get system store and so on? 🙏

**Reply by rijkvanzanten** [reply]
Changed? https://docs.directus.io/extensions/panels.html#accessing-internal-systems is still accurate from what I can tell 🙂
